packages/clerk-js/src/core/clerk.ts

```diff
@@ -1345,4 +1345,18 @@ export default class Clerk implements ClerkInterface {
       throw new Error('ClerkJS components are not ready yet.');
     }
   }
+
+  redirectFAPIInitiatedFlowIfSignedIn = async () => {
+    const redirectUrl = new URLSearchParams(window.location.search).get('redirect_url');
+    if (
+      this.instanceType !== 'production' &&
+      redirectUrl !== null &&
+      isRedirectForFAPIInitiatedFlow(this.frontendApi, redirectUrl)
+    ) {
+      if (this.session) {
```

Comment:
❓ why do we also check the `this.session` ? Where do we need it?

Reply:
If a session is present, does it mean the user is logged in?

For the OAuth 2 IDP flow in particular, if we redirect without an active session, FAPI oauth server will redirect the user back to the sign in URL to sign in, but clerkJS will redirect back to FAPI  `/oauth/authorize` and so on. We end up with an infinite loop of redirects (sign in URL -> `/oauth/authorize` -> sign in URL -> ...). 

Do other flows (magic link, ticket) that require this redirect with dvb query param need presence of a session? If not, we may need to differentiate the logic between oauth idp and the others.

Reply:
> If a session is present, does it mean the user is logged in?
yes
I agree with @mzhong9723 that we should redirect only the user is signedIn to avoid an infinite redirect loop.

> Do other flows (magic link, ticket) that require this redirect with dvb query param need presence of a session? If not, we may need to differentiate the logic between oauth idp and the others.
cc: @SokratisVidros 

Finally we should check if session is populated when this `redirectFAPIInitiatedFlowIfSignedIn` is called ?

Reply:
In ticket and magic links flows, there is no session yet. So redirection can't be conditional. We can stop the redirection flow in FAPI in the two following ways:

1. If the referrer is `sign-in-url`.
2. If we introduce another `__clerk_synced` parameter, ClerkJS will also add to the replayed FAPI URL when it passes from redirectWithAuth.

Reply:
@SokratisVidros Could you expand a bit on 2? How would it work for OAuth2 idp vs. ticket and magic link flows?

Reply:
Basically, my suggestion is to add `__clerk_fapi_replayed=true` parameter [here](https://github.com/clerkinc/javascript/blob/d53432e4a57df64cd0556d53ff20eb0a23f4d3f5/packages/clerk-js/src/utils/devBrowser.ts#L14). Then on FAPI if no session exist, but the `__clerk_fapi_replayed` is present we continue the flow without redirecting again to SignInURL for Clerk.js.

Let me know if this helps.

Reply:
Updated to use the first solution. We will check if the referrer is the sign in url. If it is and there is no session, we will not redirect and proceed so the user can sign in.